Chị cho em chỗ test case test_uses_correct_template, em đã so với đúng tên template mà nó cứ báo lỗi AssertionError: No templates used to render the response. Chỗ em comment ạ. chị check cho em.

--- a/ecommerce/tests/factories.py
+++ b/ecommerce/tests/factories.py
@@ -1,7 +1,8 @@
 from django.contrib.auth.models import User
 import factory
 from faker import Factory
-from ecommerce.models import Order, Category, Product
+from ecommerce import models
+from ecommerce.models import Cart, Order, Category, Product, Profile
 
 faker = Factory.create()
 
@@ -9,8 +10,8 @@ class UserFactory(factory.django.DjangoModelFactory):
     class Meta:
         model = User
 
-    username = factory.LazyAttribute(lambda t: faker.name())
-    password = factory.PostGenerationMethodCall('set_password', 'mysecret')
+    username = factory.Sequence(lambda n: f'user_{n}')
+    password =  'mysecret'
 
 class OrderFactory(factory.django.DjangoModelFactory):
     class Meta:
@@ -37,3 +38,19 @@ class ProductFactory(factory.django.DjangoModelFactory):
     description=faker.text()
     price=faker.random_number()
     quantity=faker.random_number()
+
+class ProfileFactory(factory.django.DjangoModelFactory):
+    class Meta:
+        model = Profile
+        django_get_or_create = ('user',)
+    
+    user=factory.SubFactory(UserFactory)
+    birthday=faker.date_object()
+    address=faker.address()
+
+class CartFactory(factory.django.DjangoModelFactory):
+    class Meta:
+        model = Cart
+
+    quantity = 1
+    product=factory.SubFactory(ProductFactory)

--- a/ecommerce/tests/test_forms.py
+++ b/ecommerce/tests/test_forms.py
@@ -1,0 +1,44 @@
+from ecommerce.models import Profile
+from django.test import TestCase
+
+import datetime
+from ..forms import ProfileForm, UserForm
+
+
+class ProfileFormTest(TestCase):
+    def test_birthday_too_far_in_future(self):
+        """Test form is invalid if birthday more than from today."""
+        date = datetime.date.today() + datetime.timedelta(days=1)
+        form = ProfileForm(data={'birthday': date})
+        self.assertFalse(form.is_valid())
+
+    def test_birthday_field_label(self):
+        """Test birthday label is 'birthday'."""
+        form = ProfileForm()
+        self.assertTrue(
+            form.fields['birthday'].label is None or
+            form.fields['birthday'].label == 'birthday')
+
+    def test_address_field_label(self):
+        """Test address label is 'address'."""
+        form = ProfileForm()
+        self.assertTrue(
+            form.fields['address'].label is None or
+            form.fields['address'].label == 'address')
+    
+    def test_address_required_fields(self):
+        form = ProfileForm(data={'birthday': '', 'address': ''})
+        self.assertFalse(form.is_valid())
+
+class UserFormTest(TestCase):
+
+    def test_username_field_label(self):
+        """Test username label is 'username'."""
+        form = UserForm()
+        self.assertTrue(
+            form.fields['username'].label is None or
+            form.fields['username'].label == 'Username')
+            
+    def test_address_required_fields(self):
+        form = ProfileForm(data={'username': '', 'first_name': '','last_name':'','email':''})
+        self.assertFalse(form.is_valid())

--- a/ecommerce/tests/test_views.py
+++ b/ecommerce/tests/test_views.py
@@ -146,6 +146,7 @@ class CartCheckoutViewTest(TestCase):
 
         self.test_product = ProductFactory(quantity='0')
         self.test_cart = CartFactory(product=self.test_product, quantity='1')
+        self.test_order = OrderFactory(user=test_user)
            
     def test_view_url_accessible_by_name(self):
         response = self.client.get(reverse('cart_checkout'))
@@ -169,19 +170,28 @@ class CartCheckoutViewTest(TestCase):
         # Check we used correct template
         self.assertTemplateUsed(response, 'ecommerce/checkout.html')
 
-    def test_product_quantity_validate(self):
+    def test_product_out_of_stock_validate(self):
         self.client.login(username = 'testuser', password = '1X<ISRUkw+tuK')
         self.client.get(reverse('cart_add', kwargs={'pk': self.test_product.pk}))
-        response = self.client.get(reverse('cart_checkout'))
-        print(response)
-        self.assertEqual(response.status_code, 200)
-        # self.assertRedirects(response, 'ecommerce/checkout.html')
+        response = self.client.get('/ecommerce/cart/checkout')
+        self.assertRedirects(response, '/ecommerce/cart' , status_code=404, msg_prefix=f"{self.test_product.product_name} is out of stock limit, please decrease your product items in your cart")
 
     def test_order_infomation_null(self):
         self.client.login(username = 'testuser', password = '1X<ISRUkw+tuK')
-        self.test_order = OrderFactory(shipping_address='',phone_number='')
-        response = self.client.get(reverse('cart_checkout'))
-        self.assertRedirects(response, 'ecommerce/checkout.html')
+        response = self.client.get('/ecommerce/cart/checkout')
+        self.assertRedirects(response,'/ecommerce/cart', status_code=404, msg_prefix="Shipping address or phone number is invalid!")
+
+    def test_post_order_information_null(self):
+        self.client.login(username = 'testuser', password = '1X<ISRUkw+tuK')
+        response = self.client.post('/ecommerce/cart/checkout')
+        self.assertRedirects(response, reverse('cart_checkout') , status_code=301, msg_prefix="Shipping address or phone number is invalid!")
+    
+    def test_post_order_information_success(self):
+        self.client.login(username = 'testuser', password = '1X<ISRUkw+tuK')
+        response = self.client.post('/ecommerce/cart/checkout', data={"shipping_address": self.test_order.shipping_address, "phone_number":self.test_order.phone_number})
+        new_order = self.client.get(reverse('order_detail', kwargs={'pk': self.test_order.pk}))
+        self.assertEqual(str(new_order.context['total_price']), self.test_order.total_price)
+        self.assertRedirects(response, reverse('products') , status_code=200, msg_prefix="Order is waiting for being approved by admin! Please wait the approve message!")
 
 class CartViewTest(TestCase):
     def setUp(self):

--- a/ecommerce/tests/test_views.py
+++ b/ecommerce/tests/test_views.py
@@ -1,25 +1,26 @@
+from django.http import response
 from django.test import TestCase
-from django.contrib.auth.models import Permission
+from django.contrib.auth.models import Permission, User
 from django.urls.base import reverse
-from ecommerce.tests.factories import OrderFactory, UserFactory
+from ecommerce.tests.factories import CartFactory, OrderFactory, ProductFactory, UserFactory
 
 class CheckOrderAllListViewTest(TestCase):
     def setUp(self):
-        self.test_user1 = UserFactory(id=1)
-        self.test_user2 = UserFactory(id=2)
-        
-        self.test_user1.save()
-        self.test_user2.save()
+        test_user1 = User.objects.create_user(username='testuser1', password='1X<ISRUkw+tuK')
+        test_user1.save()
+
+        test_user2 = User.objects.create_user(username='testuser2', password='2HJ1vRV0Z&3iD')
+        test_user2.save()
 
         permission = Permission.objects.get(name='Set order as returned')
-        self.test_user2.user_permissions.add(permission)
-        self.test_user2.save()
-
-        # Create an Order object for test_user1
-        self.test_order1 = OrderFactory(user=self.test_user1)
+        test_user1.user_permissions.add(permission)
+        test_user1.save()
 
         # Create an Order object for test_user2
-        self.test_order2 = OrderFactory(user=self.test_user2)
+        self.test_order1 = OrderFactory(user=test_user1)
+
+        # Create an Order object for test_user2
+        self.test_order2 = OrderFactory(user=test_user2)
 
     def test_redirect_if_not_logged_in(self):
         response = self.client.get(reverse('all-order'))
@@ -27,18 +28,278 @@ class CheckOrderAllListViewTest(TestCase):
         self.assertTrue(response.url.startswith('/accounts/login/'))
 
     def test_order_all_url_exists(self):
-        login = self.client.login(username = self.test_user2.username, password = self.test_user2.password)
+        login = self.client.login(username='testuser1', password='1X<ISRUkw+tuK')
         response = self.client.get('/ecommerce/order/all/')
-        self.assertEqual(response.status_code, 302)
+        self.assertEqual(response.status_code, 200)
 
     def test_forbidden_if_logged_in_but_not_correct_permission(self):
-        login = self.client.login(username = self.test_user1.username, password = self.test_user1.password)
+        login = self.client.login(username='testuser2', password='2HJ1vRV0Z&3iD')
         response = self.client.get(reverse('all-order'))
-        self.assertEqual(response.status_code, 302)
+        self.assertEqual(response.status_code, 403)
 
     def test_logged_in_with_permission_another_users_order(self):
-        login = self.client.login(username = self.test_user2.username, password = self.test_user2.password)
+        login = self.client.login(username='testuser1', password='1X<ISRUkw+tuK')
         response = self.client.get(reverse('all-order'))
 
         # Check that it lets us login. We're a manager, so we can view any users order
+        self.assertEqual(response.status_code, 200)
+
+    def test_uses_correct_template(self):
+        self.client.login(username='testuser1', password='1X<ISRUkw+tuK')
+        response = self.client.get(reverse('all-order'))
+        self.assertEqual(response.status_code, 200)
+
+        # Check we used correct template
+        self.assertTemplateUsed(response, 'ecommerce/order_list_all.html')
+
+class OrderViewTest(TestCase):
+    def setUp(self):
+        test_user = User.objects.create_user(username='testuser', password='1X<ISRUkw+tuK')
+        test_user.save()
+
+    def test_view_url_exists_at_desired_location(self):
+        response = self.client.get('/ecommerce/order/')
         self.assertEqual(response.status_code, 302)
+           
+    def test_view_url_accessible_by_name(self):
+        response = self.client.get(reverse('order_get'))
+        self.assertEqual(response.status_code, 302)
+
+    def test_redirect_if_not_logged_in(self):
+        response = self.client.get(reverse('order_get'))
+        self.assertEqual(response.status_code, 302)
+        self.assertTrue(response.url.startswith('/accounts/login/'))
+
+    def test_logged_in(self):
+        login = self.client.login(username = 'testuser', password = '1X<ISRUkw+tuK')
+        response = self.client.get(reverse('order_get'))
+        self.assertEqual(response.status_code, 200)
+
+    def test_uses_correct_template(self):
+        self.client.login(username = 'testuser', password = '1X<ISRUkw+tuK')
+        response = self.client.get(reverse('order_get'))
+        self.assertEqual(response.status_code, 200)
+
+        # Check we used correct template
+        self.assertTemplateUsed(response, 'ecommerce/order.html')
+
+class OrderDetailViewTest(TestCase):
+    def setUp(self):
+        test_user = User.objects.create_user(username='testuser', password='1X<ISRUkw+tuK')
+        test_user.save()
+
+        self.test_order = OrderFactory(user=test_user)
+           
+    def test_view_url_accessible_by_name(self):
+        response = self.client.get(reverse('order_detail', kwargs={'pk': self.test_order.pk}))
+        self.assertEqual(response.status_code, 302)
+
+    def test_logged_in(self):
+        login = self.client.login(username = 'testuser', password = '1X<ISRUkw+tuK')
+        response = self.client.get(reverse('order_detail', kwargs={'pk': self.test_order.pk}))
+        self.assertEqual(response.status_code, 200)
+
+    def test_uses_correct_template(self):
+        self.client.login(username = 'testuser', password = '1X<ISRUkw+tuK')
+        response = self.client.get(reverse('order_detail', kwargs={'pk': self.test_order.pk}))
+        self.assertEqual(response.status_code, 200)
+
+        # Check we used correct template
+        self.assertTemplateUsed(response, 'ecommerce/order_detail.html')
+
+class ProductSearchViewTest(TestCase):
+    def test_view_url_exists_at_desired_location(self):
+        response = self.client.get('/ecommerce/search')
+        self.assertEqual(response.status_code, 200)
+           
+    def test_view_url_accessible_by_name(self):
+        response = self.client.get(reverse('product_search'))
+        self.assertEqual(response.status_code, 200)
+
+    def test_uses_correct_template(self):
+        response = self.client.get(reverse('product_search'))
+        self.assertEqual(response.status_code, 200)
+
+        # Check we used correct template
+        self.assertTemplateUsed(response, 'ecommerce/product_list.html')
+
+class ProductGetViewTest(TestCase):
+    def test_view_url_exists_at_desired_location(self):
+        response = self.client.get('/ecommerce/')
+        self.assertEqual(response.status_code, 200)
+           
+    def test_view_url_accessible_by_name(self):
+        response = self.client.get(reverse('products'))
+        self.assertEqual(response.status_code, 200)
+
+    def test_uses_correct_template(self):
+        response = self.client.get(reverse('products'))
+        self.assertEqual(response.status_code, 200)
+
+        # Check we used correct template
+        self.assertTemplateUsed(response, 'ecommerce/product_list.html')
+
+class CartCheckoutViewTest(TestCase):
+    def setUp(self):
+        test_user = User.objects.create_user(username='testuser', password='1X<ISRUkw+tuK')
+        test_user.save()
+
+        self.test_product = ProductFactory(quantity='0')
+        self.test_cart = CartFactory(product=self.test_product, quantity='1')
+           
+    def test_view_url_accessible_by_name(self):
+        response = self.client.get(reverse('cart_checkout'))
+        self.assertEqual(response.status_code, 302)
+
+    def test_redirect_if_not_logged_in(self):
+        response = self.client.get(reverse('cart_checkout'))
+        self.assertEqual(response.status_code, 302)
+        self.assertTrue(response.url.startswith('/accounts/login/'))
+
+    def test_logged_in(self):
+        login = self.client.login(username = 'testuser', password = '1X<ISRUkw+tuK')
+        response = self.client.get(reverse('cart_checkout'))
+        self.assertEqual(response.status_code, 200)
+
+    def test_uses_correct_template(self):
+        self.client.login(username = 'testuser', password = '1X<ISRUkw+tuK')
+        response = self.client.get(reverse('cart_checkout'))
+        self.assertEqual(response.status_code, 200)
+
+        # Check we used correct template
+        self.assertTemplateUsed(response, 'ecommerce/checkout.html')
+
+    def test_product_quantity_validate(self):
+        self.client.login(username = 'testuser', password = '1X<ISRUkw+tuK')
+        self.client.get(reverse('cart_add', kwargs={'pk': self.test_product.pk}))
+        response = self.client.get(reverse('cart_checkout'))
+        print(response)
+        self.assertEqual(response.status_code, 200)
+        # self.assertRedirects(response, 'ecommerce/checkout.html')
+
+    def test_order_infomation_null(self):
+        self.client.login(username = 'testuser', password = '1X<ISRUkw+tuK')
+        self.test_order = OrderFactory(shipping_address='',phone_number='')
+        response = self.client.get(reverse('cart_checkout'))
+        self.assertRedirects(response, 'ecommerce/checkout.html')
+
+class CartViewTest(TestCase):
+    def setUp(self):
+        test_user = User.objects.create_user(username='testuser', password='1X<ISRUkw+tuK')
+        test_user.save()
+
+    def test_view_url_exists_at_desired_location(self):
+        response = self.client.get('/ecommerce/cart/')
+        self.assertEqual(response.status_code, 302)
+           
+    def test_view_url_accessible_by_name(self):
+        response = self.client.get(reverse('cart_get'))
+        self.assertEqual(response.status_code, 302)
+
+    def test_redirect_if_not_logged_in(self):
+        response = self.client.get(reverse('cart_get'))
+        self.assertEqual(response.status_code, 302)
+        self.assertTrue(response.url.startswith('/accounts/login/'))
+
+    def test_logged_in(self):
+        login = self.client.login(username = 'testuser', password = '1X<ISRUkw+tuK')
+        response = self.client.get(reverse('cart_get'))
+        self.assertEqual(response.status_code, 200)
+
+    def test_uses_correct_template(self):
+        self.client.login(username = 'testuser', password = '1X<ISRUkw+tuK')
+        response = self.client.get(reverse('cart_get'))
+        self.assertEqual(response.status_code, 200)
+
+        # Check we used correct template
+        self.assertTemplateUsed(response, 'ecommerce/cart.html')
+
+class CartAddViewTest(TestCase):
+    def setUp(self):
+        test_user = User.objects.create_user(username='testuser', password='1X<ISRUkw+tuK')
+        test_user.save()
+
+        self.test_product = ProductFactory()
+
+    def test_view_url_exists_at_desired_location(self):
+        response = self.client.get(reverse('cart_add', kwargs={'pk': self.test_product.pk}))
+        self.assertEqual(response.status_code, 302)
+           
+    def test_view_url_accessible_by_name(self):
+        response = self.client.get(reverse('cart_add', kwargs={'pk': self.test_product.pk}))
+        self.assertEqual(response.status_code, 302)
+
+    def test_redirect_if_not_logged_in(self):
+        response = self.client.get(reverse('cart_add', kwargs={'pk': self.test_product.pk}))
+        self.assertEqual(response.status_code, 302)
+        self.assertTrue(response.url.startswith('/accounts/login/'))
+
+class ProfileViewTest(TestCase):
+    def setUp(self):
+        test_user = User.objects.create_user(username='testuser', password='1X<ISRUkw+tuK')
+        test_user.save()
+
+    def test_view_url_exists_at_desired_location(self):
+        response = self.client.get('/ecommerce/profile/')
+        self.assertEqual(response.status_code, 404)
+           
+    def test_view_url_accessible_by_name(self):
+        response = self.client.get(reverse('get-profile'))
+        self.assertEqual(response.status_code, 302)
+
+    def test_redirect_if_not_logged_in(self):
+        response = self.client.get(reverse('get-profile'))
+        self.assertEqual(response.status_code, 302)
+        self.assertTrue(response.url.startswith('/accounts/login/'))
+
+    def test_logged_in(self):
+        login = self.client.login(username = 'testuser', password = '1X<ISRUkw+tuK')
+        response = self.client.get(reverse('get-profile'))
+        self.assertEqual(response.status_code, 200)
+
+    def test_uses_correct_template(self):
+        self.client.login(username = 'testuser', password = '1X<ISRUkw+tuK')
+        response = self.client.get(reverse('get-profile'))
+        self.assertEqual(response.status_code, 200)
+
+        # Check we used correct template
+        self.assertTemplateUsed(response, 'ecommerce/profile.html')
+
+class ProfileUpdateViewTest(TestCase):
+    def setUp(self):
+        test_user1 = User.objects.create_user(username='testuser1', password='1X<ISRUkw+tuK')
+        test_user1.save()
+
+        test_user2 = User.objects.create_user(username='testuser2', password='2HJ1vRV0Z&3iD')
+        test_user2.save()
+
+        permission = Permission.objects.get(name='Set order as returned')
+        test_user1.user_permissions.add(permission)
+        test_user1.save()
+
+    def test_redirect_if_not_logged_in(self):
+        response = self.client.get(reverse('update-profile'))
+        self.assertEqual(response.status_code, 302)
+        self.assertTrue(response.url.startswith('/accounts/login/'))
+
+    def test_logged_in(self):
+        login = self.client.login(username='testuser2', password='2HJ1vRV0Z&3iD')
+        response = self.client.get(reverse('update-profile'))
+        self.assertEqual(response.status_code, 200)
+
+    def test_uses_correct_template(self):
+        login = self.client.login(username='testuser2', password='2HJ1vRV0Z&3iD')
+        response = self.client.get(reverse('update-profile'))
+        self.assertEqual(response.status_code, 200)
+
+        # Check we used correct template
+        self.assertTemplateUsed(response, 'ecommerce/profile.html')
+
+    def test_redirects_to_profile_update_on_success(self):
+        login = self.client.login(username='testuser2', password='2HJ1vRV0Z&3iD')
+        first_name, last_name, email = 'Nguyen', 'Nabi', 'nabi@email.com'
+        response = self.client.post(reverse('update-profile'),
+                                    {'first_name': first_name,
+                                     'last_name': last_name,
+                                     'email': email})
+        self.assertEqual(response.status_code, 200)

